# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
data_path = '/content/drive/MyDrive/1. IT study/Data Loves Academy/Data Analyst/10. Модуль 10/data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [ ]:
df.sort_values(by='Unit price', ascending=False).head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [ ]:
df[df['Product line'] == 'Electronic accessories'].sort_values(
    by=['Quantity', 'Unit price'],
    ascending=[False, False]
).head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


Якщо ми хочемо знайти товар, який одночасно і найбільше продавався, і є найдорожчим, то код можна написати такий:

In [ ]:
electronic = df[df['Product line'] == 'Electronic accessories']

result = electronic[
    (electronic['Quantity'] == electronic['Quantity'].max()) &
    (electronic['Unit price'] == electronic['Unit price'].max())
]

result

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating


Як бачимо, такого товару в даному випадку у нас немає. Тому у коді перед цим ми пошукали спочатку, який товар найбільше продавався, і вже серед них шукаємо з найбільшою ціною.

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [ ]:
def category_rating(val):
    if val >= 8:
        return 'High'
    elif 6 <= val < 8:
        return 'Medium'
    else:
        return 'Low'

df['Rating Category'] = df['Rating'].apply(category_rating)
df[['Rating', 'Rating Category']].head()

,Rating,Rating Category
0,9.1,High
1,9.6,High
2,7.4,Medium
3,8.4,High
4,5.3,Low


3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [ ]:
df['Rating Category'].value_counts(normalize=True) * 100

,proportion
Rating Category,
Medium,35.6
High,32.9
Low,31.5


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [ ]:
time = '11:28'

In [ ]:
time = '11:28'

def get_hour(time):
    return int(time[:2])

print(get_hour(time))

11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [ ]:
df['Hour'] = df['Time'].apply(get_hour)
df[['Time', 'Hour']].head()

,Time,Hour
0,13:08,13
1,10:29,10
2,13:23,13
3,20:33,20
4,10:37,10


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це логічно, що найбільше покупок саме в ці години?".

In [ ]:
df['Hour'].value_counts().sort_values(ascending=False).head(24)

,count
Hour,
19,113
13,103
15,102
10,101
18,93
11,90
12,89
14,83
16,77


Написала у `.head(24)` саме 24, щоб у разі цілодобового робочого дня супермаркету вивести всі значення для кожної години. Як бачимо з виведеного результату, супермаркет працює 11 годин - з 10:00 по 21:00.

In [ ]:
hour_counts = df['Hour'].value_counts().sort_index()

print(hour_counts)

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: count, dtype: int64


Найбільше покупок здійснюють о 19:00 — 113 покупок. Також досить багато покупок припадає на 13:00 (103) та 15:00 (102) і 10:00 (101). Це може бути пов'язано з обідньою перервою та завершенням робочого дня, коли люди мають більше вільного часу для покупок. Найвища кількість покупок о 19:00 є логічною, оскільки після роботи люди часто заходять до супермаркету, щоб придбати необхідні товари.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [ ]:
df.groupby('City')['Total'].mean().round(2)

,Total
City,
Mandalay,319.87
Naypyitaw,337.10
Yangon,312.35


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [ ]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max']).round(2)

,count,mean,min,max
City,,,,
Mandalay,332,319.87,18.64,1022.49
Naypyitaw,328,337.10,10.68,1042.65
Yangon,340,312.35,12.69,1039.29


**Висновки**

1. Найбільше покупок — у Yangon: 340.
Це означає, що Yangon має найбільшу кількість транзакцій, хоча середній чек там найнижчий.

2. Найвищий середній чек — у Naypyitaw: 337.10.
Тобто покупці в цьому місті в середньому витрачають більше за одну покупку.

3. Найнижчий середній чек — у Yangon: 312.35.
Попри найбільшу кількість покупок, середня сума одного чеку там нижча.

4. Максимальні чеки в усіх трьох містах досить близькі — приблизно 1022–1043. Тому можемо сказати, що потенціал для великих покупок є в кожному місті.

**Рекомендації**

- **Yangon** варто сфокусувати на збільшенні середнього чеку: використовувати акційні набори, пропозиції '2+1' та рекомендації супутніх товарів. Адже покупок тут найбільше, але середній чек найнижчий.

- **Naypyitaw** демонструє найвищий середній чек, тому тут може бути доцільно тестувати преміальні товари та дорожчі товарні набори, щоб ще більше збільшити виручку з одного клієнта.

- **Mandalay** займає проміжну позицію, тому варто окремо дослідити, які категорії товарів формують найбільший чек, і масштабувати успішні пропозиції.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [ ]:
df.to_csv('/content/drive/MyDrive/1. IT study/Data Loves Academy/Data Analyst/10. Модуль 10/data/supermarket_sales_enriched.csv', index=False)

In [ ]:
pd.read_csv('/content/drive/MyDrive/1. IT study/Data Loves Academy/Data Analyst/10. Модуль 10/data/supermarket_sales_enriched.csv', nrows=10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category,Hour
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High,13
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High,10
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium,13
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High,20
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low,10
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1,Low,18
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8,Low,14
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0,High,11
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2,Medium,17
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9,Low,13
